### DISCUSSION TITLE
CFG files:  How do I define MEMORY and a SEGMENT and initialize the segment?

### DISCUSSION THREAD

**Discussion by brad-colbert** [cc65/cc65#2406] [General]
Hello!

I'm trying to have some memory that is configured at a specific location.  As far as I understand, in the CFG file, I need to define a MEMORY entry (for the location) and a SEGMENT.  Then, in a C file I define the array with a #pragma push/pop.

I've done just that, but the memory is not initialized with my values.  To note, if I do not use the MEMORY location (but use MAIN) that the location is initialized with my values, just not where I want it.

Any thoughts?

CFG snippets:
```
MEMORY {
...
    DISPLAYLIST_MEM: file = "",           start = $4D00,                   size = $400;
    FRAMEBUFFER_MEM: file = "",           start = $5000,                   size = $2340;
}

SEGMENTS {
...
    GFX8_DL:         load = DISPLAYLIST_MEM,                      type = rw,  align = $0100,  define = yes; #,               align=$0100; # FIXME
    GFX8_CONSOLE_DL: load = DISPLAYLIST_MEM,                      type = rw,  align = $0100,  define = yes; # FIXME
    GFX9_CONSOLE_DL: load = DISPLAYLIST_MEM,                      type = rw,  align = $0100,  define = yes; # FIXME
    FRAMEBUFFER: load = FRAMEBUFFER_MEM,               type = rw,  align = $1000,  define = yes;
...
}

```
C Code snippets:
```
#pragma data-name (push,"GFX8_DL")
void graphics_8_dl = {
    DL_BLK8,
    DL_LMS(DL_GRAPHICS8), &framebuffer[0], DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8,
    DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8,
    DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8, DL_GRAPHICS8,
...
#pragma data-name (pop)
...

```
Memory:
![image](https://github.com/cc65/cc65/assets/45603525/85f204c0-1f4b-45a2-ac78-3ac7153326e2)

Any thoughts and/or help is appreciated.

**Comment by Marrin** [comment]
From the documentation [5.3 Output files](https://cc65.github.io/doc/ld65.html#toc5.3):

> Assigning an empty file name for a memory area will discard the data written to it. This is useful, if the memory area has segments assigned that are empty (for example because they are of type bss).

So this is obviously not working.  The linker places the data into the segment and that into the memory area, but the data is explicitly not saved to a file, i.e. lost/discarded.

How to deal with this depends on the target and how executable files are handled there.

**Reply by brad-colbert** [reply]
So that's what that is for :)

Thank you

**Comment by groessler** [comment]
Create a load-chunk in your exe file. See Atari documentation.

**Comment by rofl0r** [comment]
> `void graphics_8_dl = {`

i'm surprised this compiles at all. void by itself isnt a type that can be used for data.

**Reply by groessler** [reply]
It's a cc65 extension (I don't really like...)
